KEP 4447: Promote PolicyReport API to Kubernetes SIG API (#4448)
kubernetes/enhancements: keps/sig-auth/4447-promote-policyreport-sig-api/README.md

Promoting the Policy Report API will further increase adoption and signifies enhanced reliability, improved usability, and reinforces its significance within the cloud-native ecosystem.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

The Policy Report API is already being used by several projects and tools in the Kubernetes community and CNCF ecosystem, as detailed above. Moving the API to a the `github.com/kubernetes-sigs/` GitHub org will provide a permanent home for the API, along with improved visibility and colloboration.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Add `policy-report-api` as a new project under kubernetes-sigs i.e `github.com/kubernetes-sigs/policy-report-api`
- Provide guidance on building consumers and producers

> 

Reviewer: Do we want to publish official artefacts for the API?
- YAML manifest?
- OCI image of Helm chart?
- something else?

Reviewer 2: Here's what I suggest:
- Golang client set to reuse in producers and consumers
- Generated YAMLs
- API spec
- Docs

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
- Report controller, producer, or consumer implementations. The intent is to document the API and allow projects and tools to specify their own

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

In order to make Policy Report API an official Kubernetes SIG API, we propose:

* Create a new project `policy-report-api` (__name subject to change__) repository under the `kubernetes-sigs` org following the guidelines mentioned [here](https://github.com/kubernetes/community/blob/master/github-management/kubernetes-repositories.md#rules-for-new-repositories)
* Cleanup the [existing policy-report directory](https://github.com/kubernetes-sigs/wg-policy-prototypes/tree/master/policy-report) to contain only the necessary content for migration

> * Notify existing projects and tools about the new repository location and share with relevant sigs (`sig-auth`, `sig-security`, `sig-apimachinery`)

Reviewer: ```suggestion
* Notify existing projects and tools about the new repository location and share with relevant special interest groups (SIG API Machinery, SIG Auth, SIG Security)
```

This might be a good place to talk about core concepts and how they relate.
-->

* The Policy Report CRD originated with this proposal back in 2020. Refer to [Policy Report Custom Resource Definition](https://docs.google.com/document/d/1nICYLkYS1RE3gJzuHOfHeAC25QIkFZfgymFjgOzMDVw/edit#heading=h.nhx5d35q5eth).
* Presented to `sig-auth` about the proposal to make PolicyReport an official SIG API on [June 21, 2023](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/view#heading=h.ii52a0istwiv). This KEP is the outcome of that presentation.
* List of projects using the PolicyReport CRD either from `wg-policy-prototypes` or maintain their own fork. See [this comment](https://github.com/kubernetes-sigs/wg-policy-prototypes/issues/112#issuecomment-1453209660).

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

#### load on etcd

> Based on the producer and usage, it is possible to create lots of report objects. 
> For example, if a policy engine has 20 policy rules and a namespace has 1000 pods, 
> an implementation may produce 20,000 reports. This can overwhelm etcd.

Reviewer: ```suggestion
Based on the producer and usage, it is possible to create lots of report objects. 
For example, if a policy engine has 20 policy rules and a namespace has 1000 pods, 
an implementation may produce 20,000 reports. If a cluster operator deploys PolicyReport
into their cluster, using this APU can overwhelm etcd.
```
(is this a risk? We already let people deploy any CRD they like.)

Reviewer 3: It has definitely been reported as an issue for users of the API. Whether that constitutes a risk or not is a good question, but this should be highlighted _somwhere_

Reviewer 2: We have 2 projects we can link to and reference - perhaps in the docs?
- https://github.com/kyverno/reports-server
- https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-policy/98-long-term-compliance-history

Reviewer 4: This is my core concern with the concept of reporting findings via the API server.

I think events are another example of a high-volume object? One salient difference between this and events, is that events are understood to be subject to throttling/sampling. Security reports may not have the same luxury.

I like the idea of reports-server, but IMO it would need to be an expectation that all clusters have a similar scalable backend solution before reports could be reliably enabled without risking cluster stability.

Reviewer 4: I should also add that there is a difference between "users can deploy any CRD they like" and "K8s accepts using KRM/the API server this way as a valid practice", the second statement has much stronger implications around supportability.

Reviewer 5: @JimBugwadia can you comment on the reliability and performance concerns brought up here?

Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests
N/A

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>

### Graduation Criteria
We need approvals from the following stakeholders:

> [TBD]

Reviewer: Will we target this API at a release?

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

The PolicyReport API is already being used by several projects and tools in the Kubernetes community and CNCF ecosystem, as detailed above. Moving the API to a the `github.com/kubernetes-sigs/` GitHub org will provide a permanent home for the API, along with improved visibility and collaboration.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Add `policy-report-api` as a new project under kubernetes-sigs i.e `github.com/kubernetes-sigs/policy-report-api`
- Provide guidance on building consumers and producers

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

> - Report controller, producer, or consumer implementations. The intent is to document the API and allow projects and tools to specify their own

Reviewer 3: ```suggestion
- Report controller, producer, or consumer implementations. The intent is to document the API and allow projects and tools to specify their own implementation.
```

    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

> - Adopt PolicyReport as an official, in-tree Kubernetes API

Reviewer 6: I get this proposal is not about putting the API it in-tree, but I'd like to register my opposition to putting it in-tree.

From a scalability/stability point of view, reporting seems like a secondary concern compared to running an application, and better targeted for a datastore that is not serving traffic (i.e. not the cluster running the actual workloads being audited). From a security operations point of view, I’d be concerned about the veracity of this API if it was coming from a cluster that is hosting the workload being reported on. I’d really want security reporting information to be stored in a separate domain (in this case, cluster) from the source of the data. If it were in the same cluster, any Kubernetes CVE/authorization misconfiguration becomes that much more worse and calls the authenticity of report data into question. To anyone relying on reports for compliance, that becomes a real business-impacting issue. Given all that, it feels like making such an API part of Kubernetes core would naturally lead people to adopt anti-patterns in both (stability & security) cases, making the cluster a self-contained unit where source data is gathered and reports are stored.

Author: @micahhausler Yes, this proposal is not for adding the API in-tree. Instead, it is for a uniform API for reporting. I have listed in-tree as an alternative that was considered, but ruled out as policy reports are best managed as a Custom Resource. Perhaps I should clarify that.
